code/modules/ai/ai_behaviors/human_mobs/weapon_module.dm

```diff
@@ -0,0 +1,309 @@
+	///Chat lines when reloading
+	var/list/reloading_chat = list("Reloading!", "Cover me, reloading!", "Changing mag!", "Out of ammo!")
+	///Chat lines when target goes out of range
+	var/list/out_range_chat = list("Target out of range.", "Out of range.", "I lost them.", "Where'd they go?", "They're running!")
+	///Chat lines when LOS broken
+	var/list/no_los_chat = list("Target lost!", "Where'd they go?", "I lost sight of them!", "Where'd they go?", "They're running!", "Stop hiding!")
+	///Chat lines when some asshole on your team is in the way
+	var/list/friendly_blocked_chat = list("Get out of the way!", "You're in my line!", "Clear the firing lane!", "Move!", "Holding fire!", "Stop blocking me damn it!")
+	///Chat lines when target dies or is destroyed
+	var/list/dead_target_chat = list("Target down.", "Hostile down.", "Scratch one.", "I got one!", "Down for the count.", "Kill confirmed.")
+
+///Weapon stuff that happens during process
+/datum/ai_behavior/human/proc/weapon_process()
+	if(need_weapons)
+		equip_weaponry()
+
+	if(!gun)
+		return
+	var/fire_result = can_shoot_target(combat_target)
+	if(!gun_firing)
+		if(fire_result == AI_FIRE_NO_AMMO)
+			reload_gun()
+			return
+		if(fire_result != AI_FIRE_CAN_HIT)
+			return //cant shoot yet
+		if(prob(90))
+			try_speak(pick(start_fire_chat))
+		if(gun.start_fire(mob_parent, combat_target, get_turf(combat_target)) && gun.gun_firemode != GUN_FIREMODE_SEMIAUTO && gun.gun_firemode != GUN_FIREMODE_BURSTFIRE) //failed to fire or not autofire
+			gun_firing = TRUE
+		return
+
+	if(fire_result == AI_FIRE_CAN_HIT)
+		return //keep shooting, before switch for reload consideration
+
+	stop_fire()
+
+	//already firing
+	switch(fire_result)
+		if(AI_FIRE_INVALID_TARGET)
+			return //how'd you do this?
+		if(AI_FIRE_TARGET_DEAD) //todo: add chat cooldowns based on these defines
+			if(prob(75))
+				try_speak(pick(dead_target_chat))
+		if(AI_FIRE_NO_AMMO)
+			reload_gun()
+		if(AI_FIRE_OUT_OF_RANGE)
+			if(prob(50))
+				try_speak(pick(out_range_chat))
+		if(AI_FIRE_NO_LOS)
+			if(prob(50))
+				try_speak(pick(no_los_chat))
+		if(AI_FIRE_FRIENDLY_BLOCKED)
+			if(prob(50))
+				try_speak(pick(friendly_blocked_chat))
+
+/datum/ai_behavior/human/proc/equip_weaponry(datum/source)
+	SIGNAL_HANDLER
+	if(mob_parent.incapacitated() || mob_parent.lying_angle)
+		need_weapons = TRUE
+		return
+	if(!length(mob_inventory.melee_list) && !length(mob_inventory.gun_list))
+		need_weapons = TRUE
+		engagement_range = initial(engagement_range)
+		find_weapon()
+		return
+	INVOKE_ASYNC(src, PROC_REF(do_equip_weaponry))
+
+///Searches for a weapon to pickup
+/datum/ai_behavior/human/proc/find_weapon(range = 1, list/exlusion_turfs)
+	var/list/search_list = list(get_turf(mob_parent))
+	search_list |= RANGE_TURFS(range, mob_parent)
+	for(var/turf/turf AS in search_list)
+		for(var/obj/obj in turf)
+			if(!istype(obj, /obj/item/weapon))
+				continue
+			change_action(MOVING_TO_ATOM, obj, list(0, 1))
+			return
```

Comment:
Something i forsee being a problem already is AI trying to steal peoples weapons

Reply:
If you let a weaponless angry man into melee range, punch you to the floor, then they steal your weapon and shoot you to death with it... absolutely based and you deserve it for getting owned by a bot.

Reply:
I mean,, why are there weaponless AI running around? This will ONLY trigger when they have no weapon at all. You can throw a boot knife onto the floor and they'll happily pick it up and use that.
Kinda want to test how this actually goes though, as I think this is extremely rarely going to be an issue, and you can still just disarm an AI to get your weapon back